community/modules/scripts/htcondor-install/files/autoscaler.py

```diff
@@ -257,65 +249,110 @@ def scale(self):
+            **self.zoneargs,
+            instanceGroupManager=self.instance_group_manager,
+            filter=being_born_combined_filter,
+            orderBy="creationTimestamp desc"
+        )
+        respCreatingInstances = reqCreatingInstances.execute()
+
+        # Find VMs that are idle (no dynamic slots created from partitionable
+        # slots) in the MIG handled by this autoscaler
+        filter_idle_vms = classad.ExprTree(f"PartitionableSlot && NumDynamicSlots==0")
+        filter_claimed_vms = classad.ExprTree(f"PartitionableSlot && NumDynamicSlots>0")
+        filter_mig = classad.ExprTree(f"regexp(\".*/{self.instance_group_manager}$\", CloudCreatedBy)")
+        # A full list of Machine (StartD) ClassAd attributes can be found at
+        # https://htcondor.readthedocs.io/en/latest/classad-attributes/machine-classad-attributes.html
+        idle_node_ads = coll.query(htcondor.AdTypes.Startd,
+            constraint=filter_idle_vms.and_(filter_mig),
+            projection=["Machine", "CloudZone"])
+
+        NODENAME_ATTRIBUTE = "UtsnameNodename"
+        claimed_node_ads = coll.query(htcondor.AdTypes.Startd,
+            constraint=filter_claimed_vms.and_(filter_mig),
+            projection=[NODENAME_ATTRIBUTE])
+        claimed_nodes = [ ad[NODENAME_ATTRIBUTE] for ad in claimed_node_ads]
+
+        # treat OrderedDict as a set by ignoring key values; this set will
+        # contain VMs we would consider deleting, in inverse order of
+        # their readiness to join pool (creating, unhealthy, healthy+idle)
+        idle_nodes = OrderedDict()
+        try:
+            creatingInstances = respCreatingInstances["managedInstances"]
+        except KeyError:
+            creatingInstances = []
+
+        # there is potential for nodes in MIG health check "VERIFYING" state
+        # to have already joined the pool and be running jobs
+        for instance in creatingInstances:
+            self_link = instance["instance"]
+            node_name = self_link.rsplit("/", 1)[-1]
+            if node_name not in claimed_nodes:
+                idle_nodes[self_link] = "creating"
+
+        for ad in idle_node_ads:
+            node = ad["Machine"].split(".")[0]
+            zone = ad["CloudZone"]
+            self_link = "https://www.googleapis.com/compute/v1/projects/" + \
+                self.project + "/zones/" + zone + "/instances/" + node
+            # there is potential for nodes in MIG health check "VERIFYING" state
+            # to have already joined the pool and be idle; delete them last
+            if self_link in idle_nodes:
+                idle_nodes.move_to_end(self_link)
+            idle_nodes[self_link] = "idle"
+        n_idle = len(idle_nodes)
+
+        print(f"There are {n_idle} VMs being created or idle in the pool")
+        if self.debug > 1:
+            print("Listing idle nodes:")
+            pprint(idle_nodes)
+
+        # always keep size tending toward the minimum idle VMs requested
```

Comment:
nit: This function is 200 lines long. My general rule of thumb is that growing beyond 100 usually means there is an opportunity to split up. To me it feels like the real meat of the scaling starts here and below. Much (but not all) of what is above this line is collecting information to perform the scaling. 

How about creating helper functions to gather each piece of critical information?

Reply:
ACK